src/messaging/ChannelContext.cpp

```diff
@@ -0,0 +1,407 @@
+        }
+    }
+    case ChannelState::kChanneState_Ready:
+        return mReady.mSession == session;
+    default:
+        return false;
+    }
+}
+
+void ChannelContext::EnterPreparingState(const ChannelBuilder & builder)
+{
+    mState              = ChannelState::kChanneState_Preparing;
+    mPreparing.mBuilder = builder;
+
+    // TODO: Skip waiting for interface, it is not clear how to wait for a network interface for now.
+    mPreparing.mInterface = INET_NULL_INTERFACEID;
+
+    EnterAddressResolve();
+}
+
+void ChannelContext::ExitPreparingState() {}
+
+// Address resolve
+void ChannelContext::EnterAddressResolve()
+{
+    mPreparing.mState = PrepareState::kPrepareState_AddressResolving;
+
+#ifndef NDEBUG
+    // Skip address resolve if the address is provided
+    {
+        auto addr = mPreparing.mBuilder.GetHintPeerAddress();
+        if (addr.HasValue())
+        {
+            mPreparing.mAddress = addr.Value();
+            ExitAddressResolve();
+            switch (mPreparing.mBuilder.GetSessionType())
+            {
+            case ChannelBuilder::SessionType::kSession_PASE:
+                EnterPaseParingState();
+                break;
+            case ChannelBuilder::SessionType::kSession_CASE:
+                EnterCaseParingState();
+                break;
+            }
+            return;
+        }
+    }
+#endif // NDEBUG
+
+    // TODO: call mDNS Scanner::SubscribeNode after PR #4459 is ready
+    // Scanner::RegisterScannerDelegate(this)
+    // Scanner::SubscribeNode(mPreparing.mBuilder.GetPeerNodeId())
+
+    // The HandleNodeIdResolve may already have been called, recheck the state here before set up the timer
+    if (mState == ChannelState::kChanneState_Preparing && mPreparing.mState == kPrepareState_AddressResolving)
+    {
+        System::Layer * layer = mExchangeManager->GetSessionMgr()->SystemLayer();
+        layer->StartTimer(CHIP_CONFIG_NODE_ADDRESS_RESOLVE_TIMEOUT_MSECS, AddressResolveTimeout, this);
+        Retain(); // Keep the pointer in the timer
+    }
+}
+
+void ChannelContext::AddressResolveTimeout(System::Layer * aLayer, void * aAppState, System::Error aError)
+{
+    ChannelContext * me = static_cast<ChannelContext *>(aAppState);
+    me->AddressResolveTimeout();
+    me->Release();
+}
+
+void ChannelContext::AddressResolveTimeout()
+{
+    if (mState != ChannelState::kChanneState_Preparing)
+        return;
+    if (mPreparing.mState != PrepareState::kPrepareState_AddressResolving)
+        return;
+
+    ExitAddressResolve();
+    ExitPreparingState();
+    EnterFailedState(CHIP_ERROR_PEER_NODE_NOT_FOUND);
+}
+
+void ChannelContext::HandleNodeIdResolve(CHIP_ERROR error, uint64_t nodeId, const Mdns::MdnsService & address)
+{
+    switch (mState)
+    {
+    case ChannelState::kChanneState_Ready: {
+        auto peer = mExchangeManager->GetSessionMgr()->GetPeerConnectionState(mReady.mSession);
+        if (peer->GetPeerNodeId() != nodeId || mReady.mInterface != address.mInterface)
```

Comment:
When does this happen?

Reply:
This function handles mDNS events. The events will be broadcast-ed to all channel, so every channel should check if the event is matching the channel.